server/webhook.go

```diff
@@ -39,120 +39,50 @@ const (
 func (s *Server) registerWebhookRoutes(g *echo.Group) {
 	g.POST("/gitlab/:id", func(c echo.Context) error {
 		ctx := c.Request().Context()
-		body, err := io.ReadAll(c.Request().Body)
-		if err != nil {
-			return echo.NewHTTPError(http.StatusBadRequest, "Failed to read webhook request").SetInternal(err)
+
+		webhookEndpointID, body, repos, httpErr := s.validateWebhookRequest(ctx, c)
+		if httpErr != nil {
+			return httpErr
 		}
 
-		pushEvent := &gitlab.WebhookPushEvent{}
-		if err := json.Unmarshal(body, pushEvent); err != nil {
+		var pushEvent gitlab.WebhookPushEvent
+		if err := json.Unmarshal(body, &pushEvent); err != nil {
 			return echo.NewHTTPError(http.StatusBadRequest, "Malformed push event").SetInternal(err)
 		}
-
 		// This shouldn't happen as we only setup webhook to receive push event, just in case.
 		if pushEvent.ObjectKind != gitlab.WebhookPush {
 			return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Invalid webhook event type, got %s, want push", pushEvent.ObjectKind))
 		}
-
-		branch, err := parseBranchNameFromRefs(pushEvent.Ref)
-		if err != nil {
-			return echo.NewHTTPError(http.StatusBadRequest, fmt.Sprintf("Failed to parse branch name from ref: %v", pushEvent.Ref)).SetInternal(err)
+		baseVCSPushEvent := vcs.PushEvent{
+			Ref:                pushEvent.Ref,
+			RepositoryID:       strconv.Itoa(pushEvent.Project.ID),
+			RepositoryURL:      pushEvent.Project.WebURL,
+			RepositoryFullPath: pushEvent.Project.FullPath,
+			AuthorName:         pushEvent.AuthorName,
 		}
-		webhookEndpointID := c.Param("id")
 
-		repos, err := s.store.FindRepository(ctx, &api.RepositoryFind{
-			WebhookEndpointID: &webhookEndpointID,
-		})
-		if err != nil {
-			return echo.NewHTTPError(http.StatusInternalServerError, fmt.Sprintf("Failed to respond webhook event for endpoint: %v", webhookEndpointID)).SetInternal(err)
+		filteredRepos, httpErr := filterReposCommon(&baseVCSPushEvent, repos)
```

Comment:
Don't pass in baseVCSPushEvent as reference since we're not mutating its value inside filterReposCommon.